README.md
# NimbleTCodeWifi

Firmware that implements the [Toy Code (T-Code) v0.3](https://github.com/multiaxis/tcode-spec) protocol over Wifi/Websocket for the [Nimble Connectivity Module](https://shop.exploratorydevices.com/product/connectivity-module-dev-kit/), an ESP32 controller for the [NimbleStroker](https://shop.exploratorydevices.com/).

Tested with MultiFunPlayer's WebSocket output type, which connects to a Websocket Server running on the NimbleConModule and sends T-Code commands as messages to the device. The T-Code commands are converted into position values which are then sent to the NimbleStroker actuator.

To connect the NimbleConModule to your local network:
1. Long press the encoder button until you see the flashing white Wifi LED. This indicates the device is in Local AP mode.
2. Using any wifi-enabled device with a browser (computer, phone, tablet) connect to the newly created Access Point named `NimbleTCodeWifi`.
3. Upon connecting, a Captive Portal should pop up, allowing you to configure the device for your local network, entering its SSID and password.
4. After Saving the configuration, the device will reboot and connect to the network with a local IP address. A steady white LED indicates it is connected.
5. From here you can connect to the websocket server and send commands. A blue LED will turn on when a client is connected.

If the device fails to connect to the network (ie. incorrect password), it will re-enter Local AP mode, where you will need to reconnect and reattempt the configuration.

## TCode Information

- `D0` - Identity device and firmware version: `NimbleStroker_TCode_Serial_v0.3`
- `D1` - Identify TCode version: `TCode v0.4`
- `D2` - List available axes and user range preferences:
  - `L0 0 9999 Up`: **Up/down position** linear motion (default: `5000`)
    - Maps to NimbleStroker positions: -750 to 750
  - `V0 0 9999 Vibe`: **Vibration intensity** (default: `0`)
    - An oscillation is applied to the position when sent to the NimbleStroker.
    - Maps to the amplitude of oscillation: 0 to 25 (position units)
  - `A0 0 9999 Air`: **Auxilliary air in/out valve** (default `5000`)
    - Value `0000` = air-out valve (looser)
    - Value `5000` = stop valve (default)
    - Value `9999` = air-in valve (tighter)
  - `A1 0 9999 Force`: **Force command** (default `9999`)
    - Maps to NimbleStroker force command values: 0 to 1023
    - Controls the air pressure force of the actuator(?)
  - `A2 0 9999 VibSpeed`: **Vibration speed** (default: `9999`)
    - Maps to an oscillation speed for vibration: 0 to 20hz (default 20hz)

Other info:

- Sending live control values to an axis will ease to the target value over multiple frames rather than jump immediately when the difference in change is large (> 100 t-code units, or >50 position units). This is intended to protect the user and device. ([Source1](https://github.com/mnh86/NimbleTCodeSerial/blob/6ab66638b2670115e770fdee9d2ec5c7b04f9390/include/TCodeAxis.h#L217-L228), [Source2](https://github.com/mnh86/NimbleTCodeSerial/blob/6ab66638b2670115e770fdee9d2ec5c7b04f9390/src/main.cpp#L104-L111))
- Up/down position axis values that are sent to the NimbleStroker are set as (-750 to 750) instead of the full [documented range of (-1000 to 1000)](https://github.com/ExploratoryDevices/NimbleConModule/blob/31f09fbcaa068b3d7fe8d47e44ea5ed11437c852/README.md?plain=1#L30) to avoid piston damaging the actuator (slamming occurs at min/max ranges). This aligns with the same max/min values that the NimbleStroker Pendant sends to the actuator, from debug log analysis.
- [Acutuator feedback values](https://github.com/ExploratoryDevices/NimbleConModule/blob/31f09fbcaa068b3d7fe8d47e44ea5ed11437c852/README.md?plain=1#L24-L27) are not currently returned back to clients in this firmware.

## Usage

1. Install [Windows Virtual COM Port (VCP) drivers](https://github.com/mnh86/NimbleConModule/blob/feat/docs/docs/setup-guide-windows-arduino-ide1.md#install-windows-virtual-com-port-vcp-drivers) for the USB/serial connection to the module.
2. Set up [VSCode with PlatformIO](https://randomnerdtutorials.com/vs-code-platformio-ide-esp32-esp8266-arduino/)
3. Clone this repo and open the project in VSCode
4. Build and upload this program into the NimbleConModule (USB/serial)
5. Attach the NimbleConModule to the actuator (Label A)
   - Note: Pendant connection not supported
6. Long press the Encoder Dial (2 seconds) to reset Wifi configuration and enable the configuration portal.
7. Click the Encoder Dial to toggle stop/start sending commands to the actuator.

## Testing with MultiFunPlayer

1. Launch the [MultiFunPlayer](https://github.com/Yoooi0/MultiFunPlayer) application.
2. In the Output section, add a `WebSocket` device (plus sign).
3. Set the websocket uri with the device's IP address (ie. `ws://192.168.1.100/ws`).
4. Test the connection with the play button.
5. (Optional) To configure additional axes (L0, V0, A0, A1, A2) for multi script support...
6. Open the Output Configuration panel.
7. Clone the `TCode-0.3 (default)` config, name it `NimbleTCode`.
8. Enable the channels you'd like to use. ie.:
   - `L0` Up/Down (Select "Load unnamed script") - Default value: `50%`
   - `V0` Vibrate - Default value: `0%` (off state)
   - `A0` Valve - Default value: `50%` (off state)
   - `A1` Force (can rename) - Default value: `100%` (max)
   - `A2` Vibspeed (can rename) - Default value: `100%` (max)
9. See screenshots below.

![MultiFunPlayer Device Config](./docs/MFP-device-config.jpg)

![MultiFunPlayer WebSocket Device UI](./docs/MFP-ws-device-ui.jpg)

## Attributions

- <https://github.com/ExploratoryDevices/NimbleConModule> - Official NimbleConSDK
- <https://github.com/tzapu/WiFiManager> - For Wifi Captive Portal functionality
- <https://github.com/esphome/ESPAsyncWebServer> - For websocket server functionality
- <https://github.com/tyrm/nimblestroker/> - Nimblestroker examples
- <https://github.com/Dreamer2345/Arduino_TCode_Parser> - TCode library

See also [platformio.ini](./platformio.ini) for other 3rd party OSS libraries used in this project.

CHANGELOG.md
# Change log

## v0.3 - 02/28/2023
- Change: Single click toggle will also reset the actuator state when stopped (position = 0, force = max, vibration = off)

## v0.2 - 02/21/2023
- Fix: TCode output was not being sent back out to the websocket clients.

## v0.1 - 02/20/2023
- Initial PoC release
- RFC on axes spec


src/main.cpp
#include <Arduino.h>
#include <BfButton.h>
#include <SPIFFS.h>
#include <WiFiManager.h>
#include <ESPAsyncWebServer.h>
#include <millisDelay.h>
#include "NimbleTCode.h"

#define DEVICE_NAME "NimbleTCodeWifi"

WiFiManager wm;
AsyncWebServer server(80);
AsyncWebSocket ws("/ws");

#define FIRMWAREVERSION "NimbleStroker_TCode_Wifi_v0.2"

NimbleTCode nimble(FIRMWAREVERSION);

BfButton btn(BfButton::STANDALONE_DIGITAL, ENC_BUTT, true, LOW);

void pressHandler(BfButton *btn, BfButton::press_pattern_t pattern)
{
    switch (pattern)
    {
    case BfButton::SINGLE_PRESS:
        nimble.toggle();
        if (!nimble.isRunning()) nimble.resetState();
        break;

    case BfButton::LONG_PRESS:
        // Reset Wifi settings and credentials
        if (!wm.getConfigPortalActive()) {
            wm.resetSettings();
            delay(1000);
            wm.reboot();
        }
        break;
    }
}

millisDelay clientCleanupDelay;
millisDelay ledDelay;
millisDelay logDelay;

void logTimer()
{
    if (!logDelay.justFinished()) return;
    logDelay.repeat();

    nimble.printFrameState();
}

bool wifiLedState;

void updateLEDs()
{
    if (!ledDelay.justFinished()) return;
    ledDelay.repeat();

    if (wm.getConfigPortalActive()) {
        wifiLedState = (millis() % 500 > 250); // blink every 500ms
    } else {
        wifiLedState = true;
    }

    nimble.updateEncoderLEDs();
    nimble.updateHardwareLEDs();
    nimble.updateNetworkLEDs(
        (nimble.isRunning()) ? 50 : 0,
        (wifiLedState) ? 50 : 0
    );
}

size_t lastNumClients = 0;

void cleanupClients()
{
    if (!clientCleanupDelay.justFinished()) return;
    clientCleanupDelay.repeat();

    ws.cleanupClients();
    size_t tmpNumClients = ws.count();
    if (tmpNumClients == 0 && lastNumClients > 0) {
        nimble.stop();
    }
    lastNumClients = tmpNumClients;
}

void onWsEvent(AsyncWebSocket *server, AsyncWebSocketClient *client, AwsEventType type, void *arg, uint8_t *data, size_t len)
{
    switch (type)
    {
    case WS_EVT_CONNECT:
        nimble.start();
#ifdef DEBUG
        Serial.printf("ws[%s][%u] Client connection\n", server->url(), client->id());
#endif
        break;

    case WS_EVT_DISCONNECT:
#ifdef DEBUG
        Serial.printf("ws[%s][%u] Client disconnect\n", server->url(), client->id());
#endif
        break;

    case WS_EVT_ERROR:
#ifdef DEBUG
        Serial.printf("ws[%s][%u] Error(%u): %s\n", server->url(), client->id(), *((uint16_t*)arg), (char*)data);
#endif
        break;

    case WS_EVT_PONG:
#ifdef DEBUG
        Serial.printf("ws[%s][%u] Pong[%u]: %s\n", server->url(), client->id(), len, (len)?(char*)data:"");
#endif
        break;

    case WS_EVT_DATA:
        AwsFrameInfo * info = (AwsFrameInfo*)arg;
#ifdef DEBUG
        // Serial.printf("ws[%s][%u] %s data[%llu]\n", server->url(), client->id(), (info->opcode == WS_TEXT) ? "Text" : "Binary", info->len);
        // if (info->opcode == WS_TEXT) {
        //     data[len] = 0;
        //     Serial.printf("Message: %s\n", (char*)data);
        // }
#endif
        for (size_t i=0; i < info->len; i++) {
            nimble.inputByte(data[i]);
        }
        break;
    }
}

void setupNetwork()
{
    wm.setDarkMode(true);
    wm.setMinimumSignalQuality(10);
    wm.setConfigPortalBlocking(false);
    wm.setHostname(DEVICE_NAME);
    if (wm.autoConnect(DEVICE_NAME)) {
        ws.onEvent(onWsEvent);
        server.addHandler(&ws);
        server.begin();
    }
    clientCleanupDelay.start(1000);
}

void wsTCodeCallback(const String &input) {
    ws.textAll(input);
}

void setup()
{
    // Misc init
    WiFi.mode(WIFI_STA);    // force initial wifi setting for ESP32
    nimble.setMessageCallback(wsTCodeCallback);
    nimble.init();          // initialize NimbleConModule devices
    nimble.stop();
    while (!Serial);

#ifdef DEBUG
    Serial.setDebugOutput(true);
    delay(200);
    Serial.println("\nStarting");
#else
    Serial.setDebugOutput(false);
#endif

    setupNetwork();

    // Encoder button press events
    btn.onPress(pressHandler)
        .onPressFor(pressHandler, 2000);

    // Timers setup
    ledDelay.start(30);
    logDelay.start(1000);
}

void loop()
{
    btn.read();
    wm.process();
    nimble.updateActuator();
    cleanupClients();
    updateLEDs();
#ifdef DEBUG
    //logTimer();
#endif
}

include/nimbleUtils.h
#include <Arduino.h>
#include "nimbleConModule.h"

#define MAX_POSITION_DELTA 50

int16_t framePosition = 0; // next position to send to actuator (-1000 to 1000)
int16_t targetPos = 0; // position controlled via TCode
int16_t lastFramePos = 0;  // positon that was sent to the actuator last loop frame
int16_t frameForce = IDLE_FORCE; // next force to send to the actuator (0 to 1023)

#define VIBRATION_MAX_AMP 25
#define VIBRATION_MAX_SPEED 20.0 // hz

float vibrationSpeed = VIBRATION_MAX_SPEED; // hz
uint16_t vibrationAmplitude = 0; // in position units (0 to 25)
int16_t vibrationPos = 0;

#define RUN_MODE_OFF 0
#define RUN_MODE_ON 1

bool runMode = RUN_MODE_ON;

int16_t clampPositionDelta() {
    int16_t delta = framePosition - lastFramePos;
    if (delta >= 0) {
        return (delta > MAX_POSITION_DELTA) ? lastFramePos + MAX_POSITION_DELTA : framePosition;
    } else {
        return (delta < -MAX_POSITION_DELTA) ? lastFramePos - MAX_POSITION_DELTA : framePosition;
    }
}

void printNimbleState()
{
    Serial.printf("------------------\n");
    Serial.printf("   VibAmp: %5d\n", vibrationAmplitude);
    Serial.printf(" VibSpeed: %0.2f (hz)\n", vibrationSpeed);
    Serial.printf("   TarPos: %5d\n", targetPos);
    Serial.printf("      Pos: %5d\n", lastFramePos);
    Serial.printf("    Force: %5d\n", frameForce);
    Serial.printf("    AirIn: %5d\n", actuator.airIn);
    Serial.printf("   AirOut: %5d\n", actuator.airOut);
    Serial.printf("TempLimit: %s\n",  actuator.tempLimiting ? "true" : "false");
}

void ledPulse(short int pos, short int vibPos, bool isOn = true)
{
    byte ledScale = map(abs(pos), 0, ACTUATOR_MAX_POS, 1, LED_MAX_DUTY);
    byte ledState1 = 0;
    byte ledState2 = 0;
    byte vibScale = map(abs(vibPos), 0, VIBRATION_MAX_AMP, 1, LED_MAX_DUTY);
    byte vibState1 = 0;
    byte vibState2 = 0;

    if (isOn)
    {
        if (pos < 0) {
            ledState1 = ledScale;
        } else if (pos > 0) {
            ledState2 = ledScale;
        }

        if (vibPos < 0) {
            vibState1 = vibScale;
        } else if (vibPos > 0) {
            vibState2 = vibScale;
        }
    }

    ledcWrite(ENC_LED_N,  ledState1);
    ledcWrite(ENC_LED_SE, ledState1);
    ledcWrite(ENC_LED_SW, ledState1);

    ledcWrite(ENC_LED_NE, ledState2);
    ledcWrite(ENC_LED_NW, ledState2);
    ledcWrite(ENC_LED_S,  ledState2);

    ledcWrite(ENC_LED_W, vibState1);
    ledcWrite(ENC_LED_E, vibState2);
}

include/NimbleTCode.h
#include <Arduino.h>
#include <TCode.h>
#include "nimbleConModule.h"

#define MAX_POSITION_DELTA 50
#define VIBRATION_MAX_AMP 25
#define VIBRATION_MAX_SPEED 20.0 // hz

struct nimbleFrameState {
    int16_t targetPos = 0; // target position from tcode commands
    int16_t position = 0; // next position to send to actuator (-1000 to 1000)
    int16_t lastPos = 0; // previous frame's position
    int16_t force = IDLE_FORCE; // next force value to send to actuator (0 to 1023)
    int8_t air = 0; // next air state to send to actuator (-1 = air out, 0 = stop, 1 = air in)
    int16_t vibrationPos = 0; // next vibration position
};

class NimbleTCode {
    public:
        NimbleTCode(const String &firmware) { tcode = new TCode<3>(firmware); }
        ~NimbleTCode() { delete tcode; }
        void init();
        void start() { running = true; }
        void stop() { tcode->stop(); running = false; }
        void toggle() { if (running) stop(); else start(); }
        void inputByte(byte input) { tcode->inputByte(input); }
        void updateActuator();
        void updateEncoderLEDs(bool isOn = true);
        void updateHardwareLEDs();
        void updateNetworkLEDs(uint32_t bluetooth, uint32_t wifi);
        void setVibrationSpeed(float v) { vibrationSpeed = min(max(v, (float)0), (float)VIBRATION_MAX_SPEED); }
        void setVibrationAmplitude(uint16_t v) { vibrationAmplitude = min(max(v, (uint16_t)0), (uint16_t)VIBRATION_MAX_AMP); }
        void printFrameState(Print& out = Serial);
        bool isRunning() { return running; }
        void setMessageCallback(TCODE_FUNCTION_PTR_T function) { tcode->setMessageCallback(function); }

    private:
        TCode<3> *tcode;
        bool running = true;
        float vibrationSpeed = VIBRATION_MAX_SPEED; // hz
        uint16_t vibrationAmplitude = 0; // amplitude in position units (0 to 25)
        nimbleFrameState frame;

        void handleAxisChanges();
        void handleVibrationSpeedChanges();
        void handlePositionChanges();
        void handleAirChanges();
        void handleForceChanges();
        int16_t clampPositionDelta();
};

void NimbleTCode::init()
{
    initNimbleConModule();

    tcode->init();

    tcode->axisRegister("L0", F("Up")); // Up stroke position
    tcode->axisWrite("L0", 5000, ' ', 0); // 5000: midpoint
    tcode->axisEasingType("L0", EasingType::EASEINOUT);

    tcode->axisRegister("V0", F("Vibe")); // Vibration Amplitude
    tcode->axisWrite("V0", 0, ' ', 0);    // 0: vibration off
    tcode->axisEasingType("V0", EasingType::EASEINOUT);

    tcode->axisRegister("A0", F("Air")); // Air in/out valve
    tcode->axisWrite("A0", 5000, ' ', 0); // 0: air out, 5000: stop, 9999: air in

    tcode->axisRegister("A1", F("Force"));
    tcode->axisWrite("A1", 9999, ' ', 0); // 9999: max force

    tcode->axisRegister("A2", F("VibeSpeed"));
    tcode->axisWrite("A2", 9999, ' ', 0); // 9999: max vibration speed
}

void NimbleTCode::handleAxisChanges()
{
    handleVibrationSpeedChanges();
    handlePositionChanges();
    handleAirChanges();
    handleForceChanges();
}

void NimbleTCode::handleVibrationSpeedChanges()
{
    if (!tcode->axisChanged("A2")) return;
    int val = tcode->axisRead("A2");
    int maxSpeed = VIBRATION_MAX_SPEED * 100;
    int newSpeed = map(val, 0, 9999, 0, maxSpeed);
    vibrationSpeed = float(newSpeed) / 100;
}

void NimbleTCode::handlePositionChanges()
{
    if (!tcode->axisChanged("L0")) {
        int val = tcode->axisRead("L0");
        frame.targetPos = map(val, 0, 9999, -ACTUATOR_MAX_POS, ACTUATOR_MAX_POS);
    }

    if (this->tcode->axisChanged("V0")) {
        int val = tcode->axisRead("V0");
        vibrationAmplitude = map(val, 0, 9999, 0, VIBRATION_MAX_AMP);
    }

    if (vibrationAmplitude > 0 && vibrationSpeed > 0) {
        int vibSpeedMillis = 1000 / vibrationSpeed;
        int vibModMillis = millis() % vibSpeedMillis;
        float tempPos = float(vibModMillis) / vibSpeedMillis;
        int vibWaveDeg = tempPos * 360;
        frame.vibrationPos = round(sin(radians(vibWaveDeg)) * vibrationAmplitude);
    } else {
        frame.vibrationPos = 0;
    }
    // Serial.printf("A:%5d S:%0.2f P:%5d\n",
    //     vibrationAmplitude,
    //     vibrationSpeed,
    //     vibrationPos
    // );

    int targetPosTmp = frame.targetPos;
    if (frame.targetPos - vibrationAmplitude < -ACTUATOR_MAX_POS) {
        targetPosTmp = frame.targetPos + vibrationAmplitude;
    } else if (frame.targetPos + vibrationAmplitude > ACTUATOR_MAX_POS) {
        targetPosTmp = frame.targetPos - vibrationAmplitude;
    }
    frame.position = targetPosTmp + frame.vibrationPos;
}

void NimbleTCode::handleAirChanges()
{
    if (!tcode->axisChanged("A0")) return;
    int val = tcode->axisRead("A0");

    if (val < 5000) {
        frame.air = -1;
    } else if (val > 5000) {
        frame.air = 1;
    } else {
        frame.air = 0;
    }
}

void NimbleTCode::handleForceChanges()
{
    if (!tcode->axisChanged("A1")) return;
    int val = tcode->axisRead("A1");
    frame.force = map(val, 0, 9999, 0, MAX_FORCE);
}

void NimbleTCode::updateActuator()
{
    handleAxisChanges();
    // Send packet of values to the actuator when time is ready
    if (checkTimer())
    {
        if (isRunning()) {
            frame.lastPos = clampPositionDelta();
            actuator.positionCommand = frame.lastPos;
            actuator.forceCommand = frame.force;
            actuator.airIn = (frame.air > 0);
            actuator.airOut = (frame.air < 0);
        } else {
            actuator.airIn = false;
            actuator.airOut = false;
            actuator.forceCommand = IDLE_FORCE;
        }
        sendToAct();
    }

    if (readFromAct()) // Read current state from actuator.
    { // If the function returns true, the values were updated.

        // Unclear yet if any action is required when tempLimiting is occurring.
        // A comparison is needed with the Pendant behavior.
        // if (actuator.tempLimiting) {
        //     setRunMode(RUN_MODE_OFF);
        // }

        // Serial.printf("A P:%4d F:%4d T:%s\n",
        //     actuator.positionFeedback,
        //     actuator.forceFeedback,
        //     actuator.tempLimiting ? "true" : "false"
        // );
    }
}

void NimbleTCode::updateEncoderLEDs(bool isOn)
{
    int16_t pos = frame.lastPos;
    int16_t vibPos = frame.vibrationPos;

    byte ledScale = map(abs(pos), 0, ACTUATOR_MAX_POS, 1, LED_MAX_DUTY);
    byte ledState1 = 0;
    byte ledState2 = 0;
    byte vibScale = map(abs(vibPos), 0, VIBRATION_MAX_AMP, 1, LED_MAX_DUTY);
    byte vibState1 = 0;
    byte vibState2 = 0;

    if (isOn)
    {
        if (pos < 0) {
            ledState1 = ledScale;
        } else if (pos > 0) {
            ledState2 = ledScale;
        }

        if (vibPos < 0) {
            vibState1 = vibScale;
        } else if (vibPos > 0) {
            vibState2 = vibScale;
        }
    }

    ledcWrite(ENC_LED_N,  ledState1);
    ledcWrite(ENC_LED_SE, ledState1);
    ledcWrite(ENC_LED_SW, ledState1);

    ledcWrite(ENC_LED_NE, ledState2);
    ledcWrite(ENC_LED_NW, ledState2);
    ledcWrite(ENC_LED_S,  ledState2);

    ledcWrite(ENC_LED_W, vibState1);
    ledcWrite(ENC_LED_E, vibState2);
}

void NimbleTCode::updateHardwareLEDs()
{
    ledcWrite(PEND_LED, (pendant.present) ? 50 : 0);
    ledcWrite(ACT_LED, (actuator.present) ? 50 : 0);
}

void NimbleTCode::updateNetworkLEDs(uint32_t bluetooth, uint32_t wifi)
{
    ledcWrite(BT_LED, bluetooth);
    ledcWrite(WIFI_LED, wifi);
}

int16_t NimbleTCode::clampPositionDelta()
{
    int16_t delta = frame.position - frame.lastPos;
    if (delta >= 0) {
        return (delta > MAX_POSITION_DELTA) ? frame.lastPos + MAX_POSITION_DELTA : frame.position;
    } else {
        return (delta < -MAX_POSITION_DELTA) ? frame.lastPos - MAX_POSITION_DELTA : frame.position;
    }
}

void NimbleTCode::printFrameState(Print& out)
{
    out.printf("------------------\n");
    out.printf("   VibAmp: %5d\n", vibrationAmplitude);
    out.printf(" VibSpeed: %0.2f (hz)\n", vibrationSpeed);
    out.printf("   TarPos: %5d\n", frame.targetPos);
    out.printf("      Pos: %5d\n", frame.position);
    out.printf("    Force: %5d\n", actuator.forceCommand);
    out.printf("    AirIn: %s\n", actuator.airIn ? "true" : "false");
    out.printf("   AirOut: %s\n", actuator.airOut ? "true" : "false");
    out.printf("TempLimit: %s\n", actuator.tempLimiting ? "true" : "false");
}
